protocols/kad/src/query/peers/closest/disjoint.rs

```diff
@@ -0,0 +1,739 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use super::*;
+use crate::kbucket::{Key, KeyBytes};
+use libp2p_core::PeerId;
+use std::{collections::{HashMap, HashSet}, ops::{Add, Index, IndexMut}};
+use wasm_timer::Instant;
+
+/// Wraps around a set of [`ClosestPeersIter`], enforcing a disjoint discovery
+/// path per configured parallelism according to the S/Kademlia paper.
+pub struct ClosestDisjointPeersIter {
+    config: ClosestPeersIterConfig,
+
+    /// The set of wrapped [`ClosestPeersIter`].
+    iters: Vec<ClosestPeersIter>,
+
+    /// Mapping of contacted peers by their [`PeerId`] to [`PeerState`]
+    /// containing the corresponding iterator indices as well as the response
+    /// state.
+    ///
+    /// Used to track which iterator contacted which peer. See [`PeerState`]
+    /// for details.
+    contacted_peers: HashMap<PeerId, PeerState>,
+
+    /// Index of the iterator last queried.
+    last_queried: IteratorIndex,
+}
+
+impl ClosestDisjointPeersIter {
+    /// Creates a new iterator with a default configuration.
+    pub fn new<I>(target: KeyBytes, known_closest_peers: I) -> Self
+    where
+        I: IntoIterator<Item = Key<PeerId>>,
+    {
+        Self::with_config(
+            ClosestPeersIterConfig::default(),
+            target,
+            known_closest_peers,
+        )
+    }
+
+    /// Creates a new iterator with the given configuration.
+    pub fn with_config<I, T>(
+        config: ClosestPeersIterConfig,
+        target: T,
+        known_closest_peers: I,
+    ) -> Self
+    where
+        I: IntoIterator<Item = Key<PeerId>>,
+        T: Into<KeyBytes> + Clone,
+    {
+        assert!(
+            config.parallelism <= config.num_results,
+            "In order to uphold S/Kademlia's disjoint paths guarantee one \
+             needs at least one result per disjoint path (parallelism).",
+        );
```

Comment:
I have to admit that I mostly ignored the notion of quorums so far. Thanks for the reminder!

If I understand your suggestion correctly we would do something like:

```rust
            KademliaHandlerEvent::GetRecordRes {
                record,
                closer_peers,
                user_data,
            } => {
                if let Some(query) = self.queries.get_mut(&user_data) {
                    if let QueryInfo::GetRecord {
                        key, records, quorum, cache_at
                    } = &mut query.inner.info {
                        if let Some(record) = record {
                            records.push(record);
                            if records.len() == quorum.get() {
                                // THE SUGGESTION
                                query.finish_path(source);
                            }
                        } else if quorum.get() == 1 {
```

Say that all paths already yielded a peer that returned a record, the above would not finish the entire query, but only a single path.

Instead of finishing a single path, why not finish all paths that yielded a peer that returned a record once `records.len() == quorum.get()`:

```rust
for (peer, _record) in records {
    query.finish_path(peer);
}
```

In the case where all paths yielded a peer that returned a record, this would end the query. In all other cases the query would continue. The only additional requirement would be to keep track of the peer that returned a record.

What do you think?

Reply:
> Say that all paths already yielded a peer that returned a record, the above would not finish the entire query, but only a single path.

Why wouldn't it? Note that this entire block of code is executed for every record response. So "all paths already yielded a peer that returned a record" means this code was executed at least once for each path, thus `finish_path()` has been called for each path and the query is finished. What am I missing?

Reply:
Note that the conditions would of course need to change to `if records.len() >= quorum.get() {` because may now need to wait for more results than it strictly needs, if there hasn't yet been at least one record from each path.

Reply:
> Why wouldn't it? Note that this entire block of code is executed for every record response. So "all paths already yielded a peer that returned a record" means this code was executed at least once for each path, thus finish_path() has been called for each path and the query is finished. What am I missing?

Say we have quorum of `20` and number of disjoint paths of `3`. The first 6 records come from peers from path `1`. The second 6 records come from peers from path `2`. The third 7 records come from path `3`. So far we are at 19 records, thus `finish_path` has not yet been called. Now another record arrives from a peer from path `1`. `finish_path` is called for path `1` and thus finishes the path. Path `2` and `3` are still active even though we have a total of 20 records and at least one record from each path.

Does the above make sense @romanb?

Reply:
Yes, makes sense, well spotted.